### PR TITLE
Implement queue data structure

### PR DESCRIPTION
**Pull request by Coolnesss**
* Queue implemented with a linked list

### CODE REVIEW

File: stdlib/Queue.delta [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+        if (head == null) {
+            tail = null;
+        }
+
+        size--;
+        return value;
+    }
+
+    /// Access element in the front of the queue
+    func front() -> T {
```
Comment (emillaine) [inline stdlib/Queue.delta]: We want to return a pointer to the original element here instead of a separate object.

File: stdlib/Queue.delta [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+    /// Initialize an empty Queue
+    init() {
+        head = null;
+        tail = null;
+        size = 0;
+    }
+
+    /// Add to the front of the queue
+    mutating func push(value: T*) {
```
Comment (emillaine) [inline stdlib/Queue.delta]: This should take the new element by value to be consistent with Array.push.

File: stdlib/Queue.delta [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    var tail: mutable Node<T>*?;
+    var size: uint;
+
+    /// Initialize an empty Queue
+    init() {
+        head = null;
+        tail = null;
+        size = 0;
+    }
+
```
Comment (emillaine) [inline stdlib/Queue.delta:14]: A deinitializer to free the remaining nodes would be nice.

File: stdlib/Queue.delta (hunk 1 above)
Comment (emillaine) [reply stdlib/Queue.delta]: Should this function be named `first()` to be consistent with Array?

Comment (Coolnesss) [reply stdlib/Queue.delta]: The current naming matches C++ [queue](http://www.cplusplus.com/reference/queue/queue/)

File: test/stdlib/queue-tests.delta [hunk 4]
```diff
@@ -0,0 +1,38 @@
+// RUN: check_exit_status 0 %delta run -Werror %s test-support.delta
+
+
+func main() {
+    testPushPop();
+    testEmpty();
+}
+
+func testPushPop() {
```
Comment (emlai) [inline test/stdlib/queue-tests.delta]: Can you add a test that pushes after popping?

File: test/stdlib/queue-tests.delta [hunk 5]
```diff
@@ -0,0 +1,38 @@
+// RUN: check_exit_status 0 %delta run -Werror %s test-support.delta
+
+
+func main() {
+    testPushPop();
+    testEmpty();
+}
```
Comment (emillaine) [inline test/stdlib/queue-tests.delta]: Can you add tests for:
- pushing after having popped
- calling isEmpty when the queue is not empty

**Review by emillaine — CHANGES_REQUESTED** [review]

File: stdlib/Queue.delta (hunk 1 above)
Comment (emillaine) [reply stdlib/Queue.delta]: But the naming of `Array.first()` doesn't match C++ vector which has `front()`, so I think we don't want `front()` here either.

File: stdlib/Queue.delta [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+    }
+}
+
+/// Represents a single element in the queue
+class Node<T> {
+
+    var next: mutable Node<T>*?;
+    var value: T;
+
+    init(next: mutable Node<T>*, val: T) {
```
Comment (emillaine) [inline stdlib/Queue.delta]: This initializer is never used so it can be removed.

**Review by emillaine — APPROVED** [review]

**Pull request merged**